### ISSUE TITLE
[Proposal] Add `format-change` event

### ISSUE THREAD

**Issue by willrowe** [quilljs/quill#379]
I'd like to propose adding a `format-change` event that's triggered whenever a format is given a new value in the editor. This would allow for modules to be kept separate ex: what I was trying to accomplish in #373. There's no need for the keyboard module to know anything about the toolbar module. If this event was added then the toolbar module would simply need to listen for it and update the buttons accordingly.
I can put together a PR for this if it is wanted.
Any input on what data should be passed with the event would be welcome. I was thinking that it should be the name of the format and the value it was set to, possibly the range it was applied to as well?

**Comment by willrowe** [comment]
Any thoughts on this?

**Comment by brettimus** [comment]
+1 would be cool to mess around with this if you could hack it together

**Comment by jhchen** [comment]
I agree intermodule dependency is not ideal. An alternative approach is an API to provide what formats are currently active (if the user types another character, what formats will be applied to it), which may change because of moving the cursor around or applying a format with the keyboard or toolbar. The logic for calculating the former is currently in the toolbar but could be centralized so other modules do not have to reimplement it.

**Comment by willrowe** [comment]
It would be great to have a way to basically query what formats are active for a range (whether that's the currently selected one or another) that's available easily from anywhere. However, this does not solve the problem of detecting when a format has changed value. You'd need to be constantly polling the active formats to see if anything has changed, which is not very performant. I think that a combination of API methods and an events would be very powerful.

**Comment by jhchen** [comment]
Yes sounds like a combination would be better then. As a default proposal, the API could just look like:

```javascript
editor.on('format-change', function(newFormats, oldFormats, source) {});
```

**Comment by Nyalab** [comment]
> The logic for calculating the former is currently in the toolbar but could be centralized so other modules do not have to reimplement it.
:+1: for this, i am using this editor via its api, and i had to reimplement this piece of logic

**Comment by willrowe** [comment]
@Nyalab can you like to how you implemented it or throw it in a gist?

**Comment by brettimus** [comment]
just want to note that i ran into a bug in my module that could be solved by this. so... another +1!

**Comment by Nyalab** [comment]
@willrowe something like this: https://gist.github.com/Nyalab/b98ac804725b17564402, it is made with React with JSX in ES6

it is not made for getting active formats for a non-current range, but it is easy to adapt it :)

**Comment by RyanV** [comment]
@jhchen is this something that can be revisited?  

My use case here is updating the "active" state of buttons in a React toolbar when a user clicks a toolbar button, presses a hot key combo, or any other way quill#format may be called.

The problem I am having is calls to #format with a collapsed selection does not fire any events.
Having a 'format-change' event would fill that hole.

The point above on decoupling modules is an important one that still seems like an issue. If for example i had 3 different modules that reacted differently to a change in formatting, I shouldn't have to   tell each of those modules to update anytime i call #format on a collapsed selection right?

let me know if i am missing something or if filling in that missing event should be part of a new issue.

**Comment by jhchen** [comment]
There does not appear to be the drive to thoughtfully explore what the change should be.

**Comment by andrewgosselin** [comment]
This seems like a no brainer to have an event like `form-change`? I stumbled a crossed this looking for the quill forms thinking it was a given that it existed. Hopefully this can be picked up soon as it's a relatively basic feature when compared to other editors.
